### PR TITLE
Added VaultRegistry tests

### PR DESCRIPTION
**Pull request by grigorevPavel**

### CODE REVIEW

**Review by MihanixA — CHANGES_REQUESTED** [review]
please edit this spec to be compliant with YearnVaultGovernance

**Review by MihanixA — APPROVED** [review]

File: test/VaultRegistry.test.ts [hunk 1]
```diff
@@ -1,445 +1,809 @@
...
+            describe("when address doesn't conform to IVault interface (IERC165)", () => {
+                it("reverts", async () => {
+                    await withSigner(
+                        this.erc20VaultGovernance.address,
+                        async (s) => {
+                            await expect(
+                                this.subject
+                                    .connect(s)
+                                    .registerVault(
+                                        randomAddress(),
```
Comment (MihanixA) [inline test/VaultRegistry.test.ts:521]: just got an idea 
let's create 
```solidity
// SPDX-License-Identifier: MIT

pragma solidity 0.8.9;

import "@openzeppelin/contracts/utils/introspection/ERC165.sol";

contract MockERC165 is ERC165 {
    mapping(bytes4 => bool) private _allowedInterfaceIdsMap;

    function allowInterfaceId(bytes4 interfaceId) external {
        _allowedInterfaceIdsMap[interfaceId] = true;
    }

    function denyInterfaceId(bytes4 interfaceId) external {
        delete _allowedInterfaceIdsMap[interfaceId];
    }

    function supportsInterface(bytes4 interfaceId) public view override returns (bool) {
        return _allowedInterfaceIdsMap[interfaceId];
    }
}
```

File: test/VaultRegistry.test.ts [hunk 2]
```diff
@@ -1,445 +1,809 @@
...
+                await this.subject
+                    .connect(this.admin)
+                    .stageProtocolGovernance(
+                        this.newProtocolGovernance.address
+                    );
+            });
+            it("denied: any other address", async () => {
+                let randomAddr = randomAddress();
+                while (randomAddr == this.admin.address) {
+                    randomAddr = randomAddress();
```
Comment (MihanixA) [inline test/VaultRegistry.test.ts:583]: let's remove `random == this.admin.address` and pray

File: test/VaultRegistry.test.ts [hunk 3]
```diff
@@ -1,445 +1,809 @@
...
+                                this.newProtocolGovernance.address
+                            )
+                    ).to.be.reverted;
+                });
+            });
+        });
+
+        describe("edge cases", () => {
+            describe("when new ProtocolGovernance is a zero address", () => {
+                it("does not fail", async () => {
```
Comment (MihanixA) [inline test/VaultRegistry.test.ts:599]: I think it should revert with `Exceptions.ADDRESS_ZERO` here. If there is no such check than it have to be added

Comment (MihanixA) [reply test/VaultRegistry.test.ts:599]: Imagine we've committed zero address as new ProtocolGovernance
then we'd face authorization issues
```solidity
    function _isProtocolAdmin(address sender) internal view returns (bool) {
        return _protocolGovernance.isAdmin(sender);
    }
```
Hence, we'd no longer be able to update ProtocolGovernance

File: test/VaultRegistry.test.ts [hunk 4]
```diff
@@ -1,445 +1,809 @@
...
+            });
+        });
+    });
+
+    describe("#lockNft", () => {
+        it("locks NFT (disables any transfer)", async () => {
+            await this.subject.connect(this.ownerSigner).lockNft(this.nftERC20);
+            expect(await this.subject.isLocked(this.nftERC20)).to.be.equal(
+                true
+            );
```
Comment (MihanixA) [inline test/VaultRegistry.test.ts:760]: let's add smth like 
`await expect(VaultRegistry.connect(...).transferFrom(...)).to.be.revertedWith(...)`

File: test/VaultRegistry.test.ts [hunk 5]
```diff
@@ -1,445 +1,809 @@
...
+    CustomContext) {
+    before(async () => {
+        // @ts-ignore
+        await setupDefaultContext.call(this);
+        this.deploymentFixture = deployments.createFixture(
+            async (_, options?: DeployOptions) => {
+                await deployments.fixture();
+
+                const { address } = await deployments.deploy("VaultRegistry", {
+                    from: this.deployer.address,
```
Comment (MihanixA) [inline test/VaultRegistry.test.ts:47]: As I understand, deploying new VaultRegistry instead of using the one that already exists, allows to execute VaultGovernance#createVault without triggering VaultRegistry#registerVault
However, I think this behaviour is not quite obvious and deploying a new Vault contract might be a better approach
```solidity
  const { address } = await deploy("AaveVault", {
      from: deployer,
      args: [],
      log: true,
      autoMine: true,
  });
```

**Review by MihanixA — CHANGES_REQUESTED** [review]

File: test/VaultRegistry.test.ts [hunk 6]
```diff
@@ -1,445 +1,809 @@
...
+
+        describe("access control:", () => {
+            it("allowed: any address", async () => {
+                await withSigner(randomAddress(), async (s) => {
+                    await expect(this.subject.vaultsCount()).to.not.be.reverted;
+                });
+            });
+        });
+        describe("edge cases", () => {
+            describe("when new vault is registered", () => {
```
Comment (alexeuler) [inline test/VaultRegistry.test.ts:397]: Instead of having this test it'd be nice to have a proptest: #vaultsCount() == #vaults().length. Check with @MihanixA how it's done in ProtocolGovernace tests

File: test/VaultRegistry.test.ts [hunk 7]
```diff
@@ -1,445 +1,809 @@
...
+    describe("#registerVault", () => {
+        it("binds minted ERC721 NFT to Vault address and transfers minted NFT to owner specified in args", async () => {
+            let newOwner = randomAddress();
+            expect(await this.subject.balanceOf(newOwner)).to.be.equal(0);
+            await withSigner(this.erc20VaultGovernance.address, async (s) => {
+                await this.subject
+                    .connect(s)
+                    .registerVault(this.anotherERC20Vault.address, newOwner);
+            });
+            expect(await this.subject.balanceOf(newOwner)).to.be.equal(1);
```
Comment (alexeuler) [inline test/VaultRegistry.test.ts:434]: I'd put here also nftForVault check

File: test/VaultRegistry.test.ts [hunk 8]
```diff
@@ -1,445 +1,809 @@
...
+                        .registerVault(
+                            this.anotherERC20Vault.address,
+                            this.ownerSigner.address
+                        )
+                ).to.emit(this.subject, "VaultRegistered");
+            });
+        });
+
+        describe("properties", () => {
+            it("@property: minted NFT equals to vaultRegistry#vaultsCount", async () => {
```
Comment (alexeuler) [inline test/VaultRegistry.test.ts:450]: `it("@property: ...)` is a symbolic name. The actual test should use `pit("...")` function. Check Vault Governances tests

File: test/VaultRegistry.test.ts [hunk 9]
```diff
@@ -1,445 +1,809 @@
...
+                expect(
+                    await this.subject.ownerOf(
+                        (await this.subject.vaultsCount()).toNumber()
+                    )
+                ).to.be.equal(newOwner);
+            });
+        });
+
+        describe("access control:", () => {
+            it("allowed: any VaultGovernance registered in ProtocolGovernance", async () => {
```
Comment (alexeuler) [inline test/VaultRegistry.test.ts:474]: I'd rename it:
allowed: any address with CREATE_VAULT permission (btw update to the latest `main` I think here VAULT_GOVERNANCE permission is used which is deprecated)

And the test it accordingly

File: test/VaultRegistry.test.ts (hunk 1 above)
Comment (alexeuler) [reply test/VaultRegistry.test.ts:521]: Yeah I think that's a good one

File: test/VaultRegistry.test.ts [hunk 10]
```diff
@@ -1,445 +1,809 @@
...
+                                        ethers.constants.AddressZero
+                                    )
+                            ).to.be.reverted;
+                        }
+                    );
+                });
+            });
+
+            describe("when owner address is zero", () => {
+                it("reverts", async () => {
```
Comment (alexeuler) [inline test/VaultRegistry.test.ts:531]: Let's also specify the exception like @MihanixA does

File: test/VaultRegistry.test.ts (hunk 3 above)
Comment (alexeuler) [reply test/VaultRegistry.test.ts:599]: Agree

File: test/VaultRegistry.test.ts [hunk 11]
```diff
@@ -1,445 +1,809 @@
...
+                                ethers.constants.AddressZero
+                            )
+                    ).to.not.be.reverted;
+                });
+            });
+        });
+    });
+
+    describe("#commitStagedProtocolGovernance", () => {
+        it("commits staged ProtocolGovernance resets staged ProtocolGovernance and ProtocolGovernanceTimestamp", async () => {
```
Comment (alexeuler) [inline test/VaultRegistry.test.ts:613]: Let's split into 3 cases
commits
resets stagedProtocolGovernance
resets timestamp

Comment (alexeuler) [reply test/VaultRegistry.test.ts:613]: The stage part can be moved to beforeEach

File: test/VaultRegistry.test.ts [hunk 12]
```diff
@@ -1,445 +1,809 @@
...
+                await sleep(
+                    Number(await this.protocolGovernance.governanceDelay())
+                );
+                await expect(
+                    this.subject
+                        .connect(this.admin)
+                        .commitStagedProtocolGovernance()
+                ).to.not.be.reverted;
+            });
+            it("denied: any other address", async () => {
```
Comment (alexeuler) [inline test/VaultRegistry.test.ts:650]: Let's also add denied: deployer

File: test/VaultRegistry.test.ts [hunk 13]
```diff
@@ -1,445 +1,809 @@
...
+            });
+
+            describe("when called before stagedProtocolGovernanceTimestamp", () => {
+                it("reverts", async () => {
+                    await this.subject
+                        .connect(this.admin)
+                        .stageProtocolGovernance(
+                            this.newProtocolGovernance.address
+                        );
+                    await sleep(
```
Comment (alexeuler) [inline test/VaultRegistry.test.ts:688]: Let's put smth like delay - 1 here, delay / 2 is too soft

File: test/VaultRegistry.test.ts [hunk 14]
```diff
@@ -1,445 +1,809 @@
...
+
+        describe("access control:", () => {
+            it("allowed: ProtocolGovernance Admin", async () => {
+                await expect(
+                    this.subject
+                        .connect(this.admin)
+                        .adminApprove(this.ownerSigner.address, this.nftERC20)
+                ).to.not.be.reverted;
+            });
+            it("denied: any other address", async () => {
```
Comment (alexeuler) [inline test/VaultRegistry.test.ts:722]: also - denied: deployer

File: test/VaultRegistry.test.ts (hunk 4 above)
Comment (alexeuler) [reply test/VaultRegistry.test.ts:760]: Agree

File: test/VaultRegistry.test.ts [hunk 15]
```diff
@@ -1,445 +1,809 @@
...
+
+        describe("access control:", () => {
+            it("allowed: NFT owner", async () => {
+                await expect(
+                    this.subject
+                        .connect(this.ownerSigner)
+                        .lockNft(this.nftERC20)
+                ).to.not.be.reverted;
+            });
+            it("denied: any other address", async () => {
```
Comment (alexeuler) [inline test/VaultRegistry.test.ts:776]: let's add: denied: Protocol Admin

**Review by alexeuler — CHANGES_REQUESTED** [review]

File: test/VaultRegistry.test.ts [hunk 16]
```diff
@@ -1,445 +1,809 @@
...
+                        await this.subject.nftForVault(
+                            this.protocolGovernance.address
+                        )
+                    ).to.equal(0);
+                });
+            });
+        });
+    });
+
+    describe("#isLocked", () => {
```
[the remark this answers is not in the document]
Comment (grigorevPavel) [reply test/VaultRegistry.test.ts:233]: we can not lock a random nft, because ownerSigner must be the owner of this particular nft. This is why I use this.nft now, but not random number

[the remark this answers is not in the document]
Comment (grigorevPavel) [reply test/VaultRegistry.test.ts:233]: There is one more problem with using random nft. VM Exception while processing transaction: reverted with reason string 'ERC721: owner query for nonexistent token'. (the nft has to be an emitted and validated number).

File: test/VaultRegistry.test.ts [hunk 17]
```diff
@@ -1,445 +1,809 @@
...
+        it("checks if token is locked (not transferable)", async () => {
+            await this.subject.connect(this.ownerSigner).lockNft(this.nftERC20);
```
[the remark this answers is not in the document]
Comment (grigorevPavel) [reply test/VaultRegistry.test.ts:235]: the same issue as stated above

File: test/VaultRegistry.test.ts [hunk 18]
```diff
@@ -1,445 +1,901 @@
...
+                        .stageProtocolGovernance(
+                            this.newProtocolGovernance.address
+                        )
+                ).to.be.revertedWith(Exceptions.FORBIDDEN);
+            });
+        });
+
+        describe("edge cases", () => {
+            describe("when new ProtocolGovernance is a zero address", () => {
+                it("reverts", async () => {
```
Comment (alexeuler) [inline test/VaultRegistry.test.ts:664]: "reverts with ${Exceptions.ADDRESS_ZERO}"

File: test/VaultRegistry.test.ts [hunk 19]
```diff
@@ -1,445 +1,901 @@
...
+        it("commits staged ProtocolGovernance", async () => {
+            await this.subject
+                .connect(this.admin)
+                .commitStagedProtocolGovernance();
+            expect(await this.subject.protocolGovernance()).to.be.equal(
+                this.newProtocolGovernance.address
+            );
+        });
+        it("resets staged ProtocolGovernanceTimestamp", async () => {
+            await this.subject
```
Comment (alexeuler) [inline test/VaultRegistry.test.ts:702]: assert timestamp was not 0 before

File: test/VaultRegistry.test.ts [hunk 20]
```diff
@@ -1,445 +1,901 @@
...
+                        this.newProtocolGovernance.address
+                    );
+            });
+        });
+        beforeEach(async () => {
+            await stagedFixture();
+            await sleep(this.governanceDelay);
+        });
+        it("commits staged ProtocolGovernance", async () => {
+            await this.subject
```
Comment (alexeuler) [inline test/VaultRegistry.test.ts:694]: assert it was not equal before

**Review by MihanixA — APPROVED** [review]

File: test/VaultRegistry.test.ts [hunk 21]
```diff
@@ -1,445 +1,901 @@
...
+import { REGISTER_VAULT } from "./library/PermissionIdsLibrary";
+import { address, pit, RUNS } from "./library/property";
+import { Arbitrary, integer } from "fast-check";
+import Exceptions from "./library/Exceptions";
+import { boolean } from "hardhat/internal/core/params/argumentTypes";
+
+type CustomContext = {
+    strategySigner: SignerWithAddress;
+    ownerSigner: SignerWithAddress;
+    newProtocolGovernance: Contract;
```
[the remark this answers is not in the document]
Comment (grigorevPavel) [reply test/VaultRegistry.test.ts:25]: It is inconvenient to make newProtocolGovernance local, because it is used in various descriptions. However it is possible to make global context smaller

**Review by MihanixA — APPROVED** [review]

File: test/library/Exceptions.ts [hunk 22]
```diff
@@ -18,4 +18,6 @@ export default class Exceptions {
     static readonly LOCK: string = "LCKD";
     static readonly INVALID_INTERFACE: string = "INVI";
     static readonly DISABLED: string = "DIS";
+    static readonly UNEXISTING_TOKEN: string =
```
[the remark this answers is not in the document]
Comment (MihanixA) [reply test/library/Exceptions.ts:21]: unresolving this comment
seems like it wasn't merged :)

File: test/library/Constants.ts [hunk 23]
```diff
@@ -2,3 +2,4 @@ export const ERC165_INTERFACE_ID: string = "0x01ffc9a7";
 export const CHIEF_TRADER_INTERFACE_ID: string = "0x698afc85";
 export const TRADER_INTERFACE_ID: string = "0xdf1e4f02";
 export const ZERO_INTERFACE_ID: string = "0x00000000";
+export const VAULT_INTERFACE: string = "0x13ab899c";
```
Comment (MihanixA) [inline test/library/Constants.ts:5]: lets use `VAULT_INTERFACE_ID` to be consistent with other variables

File: contracts/VaultRegistry.sol [hunk 24]
```diff
@@ -89,6 +92,7 @@ contract VaultRegistry is IVaultRegistry, ERC721 {
     /// @inheritdoc IVaultRegistry
     function stageProtocolGovernance(IProtocolGovernance newProtocolGovernance) external {
         require(_isProtocolAdmin(msg.sender), ExceptionsLibrary.FORBIDDEN);
+        require(address(newProtocolGovernance) != address(0), ExceptionsLibrary.ADDRESS_ZERO);
```
Comment (MihanixA) [inline contracts/VaultRegistry.sol:95]: 👍

File: contracts/VaultRegistry.sol [hunk 25]
```diff
@@ -101,6 +105,7 @@ contract VaultRegistry is IVaultRegistry, ERC721 {
         require(block.timestamp >= _stagedProtocolGovernanceTimestamp, ExceptionsLibrary.TIMESTAMP);
         _protocolGovernance = _stagedProtocolGovernance;
         delete _stagedProtocolGovernanceTimestamp;
+        delete _stagedProtocolGovernance;
```
Comment (MihanixA) [inline contracts/VaultRegistry.sol:108]: 👍

File: test/library/Exceptions.ts (hunk 22 above)
[the remark this answers is not in the document]
Comment (grigorevPavel) [reply test/library/Exceptions.ts:21]: we decided to leave this as is

File: test/VaultRegistry.test.ts [hunk 26]
```diff
@@ -1,445 +1,943 @@
...
+                expect(await this.subject.vaultForNft(this.nft)).to.equal(
+                    this.erc165Mock.address
+                );
+                expect(await this.subject.vaultForNft(this.randomNft)).to.equal(
+                    ethers.constants.AddressZero
+                );
+                await this.anotherERC165Mock.allowInterfaceId(
+                    VAULT_INTERFACE_ID
+                );
+                this.randomNft = Number(
```
Comment (alexeuler) [inline test/VaultRegistry.test.ts:200]: Looks like this should be a local variable

File: test/VaultRegistry.test.ts [hunk 27]
```diff
@@ -1,445 +1,943 @@
...
+                            this.subject.connect(s).vaultForNft(this.nft)
+                        ).to.not.be.reverted;
+                    });
+                });
+            });
+
+            describe("edge cases", () => {
+                describe("when Vault NFT is not registered in VaultRegistry", () => {
+                    it("returns zero address", async () => {
+                        let randomNumber = Math.round(Math.random() * 100 + 10);
```
Comment (alexeuler) [inline test/VaultRegistry.test.ts:232]: I thought that was the purpose of having randomNft - not to regenerate this each time

File: test/VaultRegistry.test.ts (hunk 26 above)
Comment (alexeuler) [reply test/VaultRegistry.test.ts:200]: Reassigning to this NFT would defeat the purpose of it being random and break the semantics of other test cases

Comment (alexeuler) [reply test/VaultRegistry.test.ts:200]: To avoid these kind of situations I'd rather make a helper randomNft = () => Math.random(...), use it everywhere in test cases and remove randomNft from context

File: test/VaultRegistry.test.ts [hunk 28]
```diff
@@ -1,445 +1,943 @@
...
+import { integer } from "fast-check";
+import Exceptions from "./library/Exceptions";
+import { VAULT_INTERFACE_ID } from "./library/Constants";
+import { contract } from "./library/setup";
+
+type CustomContext = {
+    ownerSigner: SignerWithAddress;
+    newProtocolGovernance: Contract;
+    allowedRegisterVaultSigner: SignerWithAddress;
+    erc165Mock: Contract;
```
Comment (alexeuler) [inline test/VaultRegistry.test.ts:26]: Since this actually mocks a vault, I'd name it vaultMock

File: test/VaultRegistry.test.ts [hunk 29]
```diff
@@ -1,445 +1,943 @@
...
+                });
+            });
+        });
+
+        describe("#stagedProtocolGovernance", () => {
+            it("returns ProtocolGovernance address staged for commit", async () => {
+                await this.subject
+                    .connect(this.admin)
+                    .stageProtocolGovernance(
+                        this.newProtocolGovernance.address
```
Comment (alexeuler) [inline test/VaultRegistry.test.ts:318]: can we replace it with randomAddress() ?

File: test/VaultRegistry.test.ts [hunk 30]
```diff
@@ -1,445 +1,943 @@
...
+                    await withSigner(randomAddress(), async (s) => {
+                        await expect(this.subject.connect(s).vaultsCount()).to
+                            .not.be.reverted;
+                    });
+                });
+            });
+
+            describe("properties", () => {
+                pit(
+                    "@property: when N new vaults have been registered, vaults count will be increased by N",
```
Comment (alexeuler) [inline test/VaultRegistry.test.ts:434]: no need to add @property - it's added by `pit` helper

File: test/VaultRegistry.test.ts [hunk 31]
```diff
@@ -1,445 +1,943 @@
...
+                    { numRuns: RUNS.verylow },
+                    integer({ min: 1, max: 5 }),
```
Comment (alexeuler) [inline test/VaultRegistry.test.ts:436]: I'd make min: 0

File: test/VaultRegistry.test.ts [hunk 32]
```diff
@@ -1,445 +1,943 @@
...
+                                );
+                        }
+                        let newVaultsCount = Number(
+                            await this.subject.vaultsCount()
+                        );
+                        expect(newVaultsCount - oldVaultsCount).to.be.equal(
+                            vaultsCount
+                        );
+                        expect(
+                            (await this.subject.vaults()).length
```
Comment (alexeuler) [inline test/VaultRegistry.test.ts:467]: pls check that if you use (await this.subject.vaults()).length + 1 here - the test will fail

File: test/VaultRegistry.test.ts [hunk 33]
```diff
@@ -1,445 +1,943 @@
...
+                        ).to.be.equal(await this.subject.vaultsCount());
+                        return true;
+                    }
+                );
+            });
+
+            describe("edge cases", () => {
+                describe("when new vault is registered", () => {
```
Comment (alexeuler) [inline test/VaultRegistry.test.ts:475]: I think this one is covered by the prop test, could remove that

File: test/VaultRegistry.test.ts [hunk 34]
```diff
@@ -1,445 +1,943 @@
...
+                pit(
+                    "@property: when N new vaults have been registered, vaults count will be increased by N",
+                    { numRuns: RUNS.verylow },
+                    integer({ min: 1, max: 5 }),
+                    async (vaultsCount: number): Promise<boolean> => {
+                        let oldVaultsCount = Number(
+                            await this.subject.vaultsCount()
+                        );
+                        for (var i = 0; i < vaultsCount; ++i) {
+                            await deployments.deploy("MockERC165", {
```
Comment (alexeuler) [inline test/VaultRegistry.test.ts:442]: deployment takes a considerable time. I'd make a fixture here of deploying 5 Mock contracts. Inside the cycle just call registerVault

File: test/VaultRegistry.test.ts [hunk 35]
```diff
@@ -1,445 +1,943 @@
...
+                            this.anotherERC165Mock.address,
+                            this.ownerSigner.address
+                        )
+                ).to.emit(this.subject, "VaultRegistered");
+            });
+
+            describe("properties", () => {
+                pit(
+                    "@property: minted NFT equals to vaultRegistry#vaultsCount",
+                    { numRuns: 1 },
```
Comment (alexeuler) [inline test/VaultRegistry.test.ts:533]: numRuns: mid, 1 is never a good idea

File: test/VaultRegistry.test.ts [hunk 36]
```diff
@@ -1,445 +1,943 @@
...
+                        expect(
+                            Number(await this.subject.vaultsCount())
+                        ).to.be.equal(Number(newNft));
+                        return true;
+                    }
+                );
+            });
+
+            describe("access control:", () => {
+                it("allowed: any account with Register Vault permissions", async () => {
```
Comment (alexeuler) [inline test/VaultRegistry.test.ts:557]: I think we also need to add a test case for ProtocolGovernance admin

File: test/VaultRegistry.test.ts [hunk 37]
```diff
@@ -1,445 +1,943 @@
...
+                                )
+                        ).to.be.revertedWith(Exceptions.FORBIDDEN);
+                    });
+                });
+            });
+
+            describe("edge cases", () => {
+                describe("when address doesn't conform to IVault interface (IERC165)", () => {
+                    it(
+                        "reverts with " + Exceptions.INVALID_INTERFACE,
```
Comment (alexeuler) [inline test/VaultRegistry.test.ts:585]: \`reverts with ${Exceptions.INVALID_INTERFACE}\` is slightly more clear

File: test/VaultRegistry.test.ts [hunk 38]
```diff
@@ -1,445 +1,943 @@
...
+                    await withSigner(randomAddr, async (s) => {
+                        await expect(
+                            this.subject.commitStagedProtocolGovernance()
+                        ).to.be.revertedWith(Exceptions.FORBIDDEN);
+                    });
+                });
+            });
+
+            describe("edge cases", () => {
+                describe("when nothing staged", () => {
```
Comment (alexeuler) [inline test/VaultRegistry.test.ts:764]: when nothing is staged

File: test/VaultRegistry.test.ts [hunk 39]
```diff
@@ -1,445 +1,943 @@
...
+            it("approves token to new address", async () => {
+                let randomAddr = randomAddress();
+                await this.subject
+                    .connect(this.admin)
+                    .adminApprove(randomAddr, this.nft);
+                expect(await this.subject.getApproved(this.nft)).to.be.equal(
+                    randomAddr
+                );
+            });
+
```
Comment (alexeuler) [inline test/VaultRegistry.test.ts:819]: Let's also add a case "clears approve from previous address"

**Review by alexeuler — CHANGES_REQUESTED** [review]

**Comment by alexeuler** [comment]
[VaultRegistry tests](https://trello.com/c/LWd8VqNh/262-vaultregistry-tests)

File: test/VaultRegistry.test.ts (hunk 29 above)
Comment (grigorevPavel) [reply test/VaultRegistry.test.ts:308]: with current functionality it will work

**Review by alexeuler — APPROVED** [review]
